### ISSUE TITLE
TaskLocal.bind might not restore value on cancelled tasks

### ISSUE THREAD

**Issue by oleg-py** [monix/monix#601]
If task passed to `bind` is cancelled, it is possible for `TaskLocal` to never be restored to its original state:
```scala
import monix.eval._
import scala.concurrent.duration._

object Sample extends TaskApp {
  override def runc: Task[Unit] = for {
    local <- TaskLocal[String]("Good")
    forked <- Task.sleep(1.second).fork
    _ <- forked.cancel
    _ <- local.bind("Bad!")(forked).fork
    _ <- Task.sleep(1.second)
    s <- local.read
    _ <- Task.eval(println(s)) // prints Bad!
  } yield ()
}
```

This can probably be solved by using `bracket` to implement `TaskLocal.bind`

**Comment by alexandru** [comment]
Indeed, we might need to implement it in terms of `bracket`.

**Comment by leandrob13** [comment]
@oleg-py  I can work on this if you haven't started it already

**Comment by leandrob13** [comment]
@oleg-py I couldn't reproduce the error in your example. I changed it to this:

```scala
implicit val ec: Scheduler = monix.execution.Scheduler.Implicits.global
implicit val opts = Task.defaultOptions.enableLocalContextPropagation

val test: Task[Unit] = for {
      local <- TaskLocal[String]("Good")
      forked <- Task.sleep(1.second).fork
      //_ <- forked.cancel
      _ <- local.bind("Bad!")(forked.cancel).fork // needed a Task and forked is a Fiber
      _ <- Task.sleep(1.second)
      s <- local.read
      _ <- Task.now(println(s)) // prints Good
    } yield ()

test.runAsyncOpt
```

And I am getting `Good` printed.

I also changed it to this:

```scala
val test: Task[Unit] = for {
      local <- TaskLocal[String]("Good")
      forked <- Task.sleep(1.second).fork
      //_ <- forked.cancel
      _ <- local.bind("Bad!")(forked.cancel) // needed a Task and forked is a Fiber
      _ <- Task.sleep(1.second)
      s <- local.read
      _ <- Task.now(println(s)) // prints Good
    } yield ()
```

So it would stay on the same `Thread` after the bind and still got `Good` printed.

Am I missing something?

How can we assert that in fact a `Task` was cancelled this way?

**Comment by oleg-py** [comment]
@leandrob13 I might have gotten this completely wrong if it works properly with `runAsyncOpt` on all scheduler types. Could be that it is not a problem.

> `local.bind("Bad!")(forked.cancel).fork // needed a Task and forked is a Fiber`

That is supposed to be `forked.join`. `cancel` terminates, and if the `Task` terminates, there would not be a problem anyway.

> How can we assert that in fact a Task was cancelled this way?

That's one of the things `bracket` does. If you do `Task.unit.bracket(_ => maybeCancelled)(_ => release)`, then `release` will be always executed, even if `maybeCancelled` is cancelled.

**Comment by leandrob13** [comment]
@oleg-py Either way I think your warning is legit because bracket handles success, failure and cancellation which is much better. I will test again changing your observations and see how it behaves with the current bind. Thanks for clarifying, I am not well familiarized with `Fiber` yet.

**Comment by alexandru** [comment]
@oleg-py I merged @leandrob13's PR #603

If you have time, please confirm that this is fixed so we can close the issue.

**Comment by oleg-py** [comment]
@alexandru looks good to me. `TaskLocal` seems to work properly in all cases I could think of, as long as I local context propagation is enabled.
